lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+# You may have seen references to v0 and v1 of our auth cookie in the codebase
+# and you're not sure how they differ, so here is an explanation:
+#
+# From the very early days of Discourse, the auth cookie (_t) consisted only of
+# a 32 characters random string that Discourse used to identify/lookup the
+# current user. We didn't include any metadata with the cookie or encrypt/sign
+# it.
+#
+# That was v0 of the auth cookie until Nov 2021 when we merged a change that
+# required us to store additional metadata with the cookie so we could get more
+# information about current user early in the request lifecycle before we
+# performed database lookup. We also started encrypting and signing the cookie
+# to prevent tampering and obfuscate user information that we include in the
+# cookie. This is v1 of our auth cookie and we still use it to this date.
+#
+# We still accept v0 of the auth cookie to keep users logged in, but upon
+# cookie rotation (which happen every 10 minutes) they'll be switched over to
+# the v1 format.
+#
+# We'll drop support for v0 after Discourse 2.9 is released.
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        db_name = RailsMultisite::ConnectionManagement.current_db
+        encryption_key = Rails.application.key_generator.generate_key(
+          "discourse-auth-cookie-encryption-#{db_name}",
+          32
+        )
+        sign_key = Rails.application.key_generator.generate_key(
+          "discourse-auth-cookie-sign-#{db_name}",
+          32
+        )
+        ActiveSupport::MessageEncryptor.new(
+          encryption_key,
+          sign_key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
```

Comment:
Why use `aes-256-cbc` here? I believe that the current [rails default](https://github.com/rails/rails/blob/037339e892e4f419e0780554f3b6fd5c0f6134fd/actionpack/lib/action_dispatch/middleware/cookies.rb#L640-L643) is `aes-256-gcm`.

I don't claim to understand the difference... but if we're using something non-default, we should be extremely careful.

The more I think about this, the more I wonder if we should just lean on Rails entirely. That way, we don't have to worry about keeping this up-to-date as Rails changes the cookie encryption method. It looks like they [automatically take care of migrations](https://github.com/rails/rails/blob/037339e892e4f419e0780554f3b6fd5c0f6134fd/actionpack/lib/action_dispatch/middleware/cookies.rb#L640-L643) whenever they change the encryption scheme.

`DefaultCurrentUserProvider` could do something like this:

```ruby
request = ActionDispatch::Request.new(env)

request.cookie_jar.encrypted[:_t] = {
  value: { user_id: 1, issued_at: Time.zone.now, trust_level: 2, token: "abcde" },
  httponly: true,
  secure: SiteSetting.force_https
}
```
and then to read it, simply:
```ruby
request = ActionDispatch::Request.new(env)
token_data = request.cookie_jar.encrypted[:_t] # Will return a hash if valid. `nil` if invalid
```
Rails then takes care of all the encryption/signing/keys/validation/etc. in the background.

What do you think @OsamaSayegh @tgxworld?

Reply:
I don't think rails default is gcm, the `request.use_authenticated_cookie_encryption` condition is tied to the `action_dispatch.use_authenticated_cookie_encryption` app config which is false by default (i.e. the else branch in the code you linked is executed by default):

https://github.com/rails/rails/blob/25f6a8e84fb6c27f489fc3e593656ddb72f25613/actionpack/lib/action_dispatch/railtie.rb#L23

Also confirmed locally in the rails console:

```
[1] pry(main)> Rails.application.config.action_dispatch.use_authenticated_cookie_encryption
=> false
```

Anyways, you do raise an excellent point about relying on `ActionDispatch::Request` to do the encryption for us. Will see if I can get it to work and confirm we're not relying on private rails APIs 👍

Reply:
According to the [API](https://api.rubyonrails.org/v6.1/classes/ActionDispatch/Cookies.html), I think we can just do `cookies.encrypted[TOKEN_COOKIE]`? That means we do not need to handle the encryption in this class, we can just make sure we set the cookies via `cookies.encrypted`.

Reply:
I wish it was that simple 😅 we need to read/decrypt our cookie in the request tracker middleware that runs so early we don't have access to Rails controller methods/helpers and that includes `cookies.encrypted`. But yeah I do think we can simplify/rely on Rails more and I'm working on that right now!

Reply:
ok @davidtaylorhq @tgxworld I've made this change. let me know what you think.